src/core_plugins/kibana/public/context/app.js

```diff
@@ -0,0 +1,103 @@
+import _ from 'lodash';
+
+import uiModules from 'ui/modules';
+import contextAppTemplate from './app.html';
+import './components/loading_button';
+import './components/size_picker/size_picker';
+import { bindSelectors } from './utils/selectors';
+import {
+  QueryParameterActionsProvider,
+  QUERY_PARAMETER_KEYS,
+} from './query_parameters';
+import {
+  QueryActionsProvider,
+  selectIsLoadingAnchorRow,
+  selectIsLoadingPredecessorRows,
+  selectIsLoadingSuccessorRows,
+  selectRows,
+} from './query';
+
+const module = uiModules.get('apps/context', [
+  'kibana',
+  'ngRoute',
+]);
+
+module.directive('contextApp', function ContextApp() {
+  return {
+    bindToController: true,
+    controller: ContextAppController,
+    controllerAs: 'contextApp',
+    restrict: 'E',
+    scope: {
+      anchorUid: '=',
+      columns: '=',
+      indexPattern: '=',
+      predecessorCount: '=',
+      successorCount: '=',
+      sort: '=',
+    },
+    template: contextAppTemplate,
+  };
+});
+
+function ContextAppController($scope, Private) {
+  const queryParameterActions = Private(QueryParameterActionsProvider);
+  const queryActions = Private(QueryActionsProvider);
+
+  this.state = createInitialState();
+
+  this.actions = _.mapValues({
+    ...queryParameterActions,
+    ...queryActions,
+  }, (action) => (...args) => action(this.state)(...args));
+
+  this.derivedState = bindSelectors({
+    isLoadingAnchorRow: selectIsLoadingAnchorRow,
+    isLoadingPredecessorRows: selectIsLoadingPredecessorRows,
+    isLoadingSuccessorRows: selectIsLoadingSuccessorRows,
+    rows: selectRows,
+  }, () => this.state);
+
+  /**
+   * Sync query parameters to arguments
+   */
+  $scope.$watchCollection(
+    () => _.pick(this, QUERY_PARAMETER_KEYS),
+    (newValues) => {
+      // break the watch cycle
```

Comment:
@weltenwort this is what I'm proposing:

```
  $scope.$watch(
    () => _.pick(this, QUERY_PARAMETER_KEYS),
    (newValues) => {
      this.actions.fetchAllRowsWithNewQueryParameters(newValues);
    },
    true
  );

  $scope.$watch(
    () => this.state.queryParameters,
    (newValues) => {
      _.assign(this, newValues);
    },
    true
  );
```

To provide an example, let's say something in `this.state.queryParameters` changes. 

1. That will fire `_.assign(this, newValues);`.
2. This change will be detected by the first watch and fire `this.actions.fetchAllRowsWithNewQueryParameters(newValues);`
3. `fetchAllRowsWithNewQueryParameters` will assign a new object to `this.state.queryParameters`, but the object property values will be the same.
4. Since we enabled deep equality checking in the second watch, it will not fire its listener because the properties of `this.state.queryParameters` haven't changed.